vllm/v1/sample/logits_processor.py

```diff
@@ -0,0 +1,242 @@
+# SPDX-License-Identifier: Apache-2.0
+import dataclasses
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Sequence, Set, Tuple
+
+import torch
+
+from vllm import SamplingParams
+
+
+@dataclasses.dataclass
+class BatchUpdate:
+    # Batch indices of any removed requests.
+    removed: List[int]
+    # (from, to) batch indices of any requests
+    # moved within the batch.
+    moved: List[Tuple[int, int]]
+    # (index, params, output_tok_ids) for new
+    # requests added to the batch.
+    added: List[Tuple[int, SamplingParams, List[int]]]
+    # The current number of requests in the batch.
+    batch_size: int
+
+
+class LogitsProcessor(ABC):
+
+    @abstractmethod
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+    @abstractmethod
+    def update_states(
+        self,
+        batch_update: Optional[BatchUpdate] = None,
+    ) -> None:
+        """Called when there are new output tokens, prior
+        to each forward pass.
+
+        Args:
+            batch_update is non-None iff there have been
+            changes to the batch makeup.
+        """
+        raise NotImplementedError
+
+
+###### ----- LogitsProcessor impls below here
+
+
+class MinPLogitsProcessor(LogitsProcessor):
+
+    def __init__(self, max_num_reqs: int, pin_memory: bool,
+                 device: torch.device):
+        self.min_p_count: int = 0
+
+        self.min_p_cpu_tensor = torch.zeros((max_num_reqs, ),
+                                            dtype=torch.float32,
+                                            device="cpu",
+                                            pin_memory=pin_memory)
+        self.min_p_cpu = self.min_p_cpu_tensor.numpy()
+        # Pre-allocated device tensor
+        self.min_p_gpu: torch.Tensor = torch.empty((max_num_reqs, ),
+                                                   dtype=torch.float32,
+                                                   device=device)
```

Comment:
Can we preallocate for other as well?

Reply:
@WoosukKwon this could potentially be done but it's not quite a simple because the maximum size could be much larger (e.g. `max_batch_size * max_logit_bias_tokens * 3` for logit bias), but they are sparse and so in practice the size would be much smaller, or more likely they wouldn't actually be used at all.

Intuitively, since we are minimizing how often these get updated, and still doing the transfer from cpu async from pinned mem, I would guess we aren't losing much by not preallocating. But I can benchmark and see. If we do it we may want to just allocate to high-water-mark so that if the feature isn't used for example then there will be no allocation.

Reply:
Got it. Sounds good. As this is a niche feature, I think we shouldn't spend too much time in it.